v2/tools/generator/internal/codegen/pipeline/state.go

```diff
@@ -131,10 +109,43 @@ func (s *State) CheckFinalState() error {
 // copy creates a new independent copy of the state
 func (s *State) copy() *State {
 	return &State{
-		definitions:        s.definitions.Copy(),
-		conversionGraph:    s.conversionGraph,
-		exportedConfigMaps: s.exportedConfigMaps.Copy(),
-		stagesSeen:         s.stagesSeen,
-		stagesExpected:     s.stagesExpected,
+		definitions:    s.definitions.Copy(),
+		stagesSeen:     s.stagesSeen,
+		stagesExpected: s.stagesExpected,
+		stateInfo:      maps.Clone(s.stateInfo),
+	}
+}
+
+// StateWithInfo returns a new state with the given information included.
+// Any existing information with the same key is replaced.
+// Has to be written as a standalone function because methods can't introduce new generic variation.
+func StateWithInfo[I any](
+	state *State,
+	key StateInfo,
+	info I,
+) *State {
+	result := state.copy()
+	if result.stateInfo == nil {
+		result.stateInfo = make(map[StateInfo]any)
 	}
+
+	result.stateInfo[key] = info
+	return result
+}
+
+// GetStateInfo returns the information stored in the state under the given key.
+// If no information is stored under the key, or if it doesn't have the expected type, the second
+// return value is false.
+// Has to be written as a standalone function because methods can't introduce new generic variation.
```

Comment:
```suggestion
// Has to be written as a standalone function because methods can't introduce new generic variables.
```
?

I wish this weren't the case, annoying gap.